### PR TITLE
feature/111 import edit history

### PR DESCRIPTION
**Pull request by thatbudakguy**
- Adds functions for generating synthetic log entries based on "input by: " and "date entered: " columns
- Adds a migration to create a generic team user for events with no clear person
- Adds a fixture to import historic geniza editors as non-active user accounts
- Adds logging to the import_data command

### CODE REVIEW

**Comment by thatbudakguy** [comment]
At the moment tests aren't passing; the error message doesn't indicate why (the objects look identical but fail comparison). I think it's a problem with the datetime objects; they were passing before...would appreciate guidance!

The actual command seems to run and do what I expect. Tests are helpful for understanding how it's supposed to work.

File: geniza/corpus/management/commands/import_data.py [hunk 1]
```diff
@@ -336,3 +372,153 @@ def log_creation(self, *objects):
...
+            dt = datetime(year=event["date"].year,
+                          month=event["date"].month,
+                          day=event["date"].day)
+            LogEntry.objects.create(
+                user=event["user"],                            # FK in django
+                content_type=self.content_types[Document],     # FK in django
+                object_id=str(doc.pk),         # TextField in django
+                object_repr=str(doc)[:200],    # CharField with limit in django
+                change_message=event["type"],
+                action_flag=ADDITION if event["type"] == "Created" else CHANGE,
```
Comment (rlskoeser) [inline geniza/corpus/management/commands/import_data.py:511]: could we just use the django flags in the event type and have th legacy input/change message keyed on that flag?

File: geniza/corpus/management/commands/import_data.py [hunk 2]
```diff
@@ -336,3 +372,153 @@ def log_creation(self, *objects):
...
+        Always creates an entry by the script user (`SCRIPT_USERNAME`) with a
+        timestamp of now to mark the import event itself.
+
+        Optional `events` is list of dict from get_edit_history. Each event has
+        a type (creation or revision), associated user, and timestamp.
+        """
+
+        # for each historic event, create a corresponding django log entry.
+        # we use objects.create() instead of the log_action helper so that we
+        # can control the timestamp.
```
Comment (rlskoeser) [inline geniza/corpus/management/commands/import_data.py:500]: 👍 thanks for documenting

File: geniza/corpus/management/commands/import_data.py [hunk 3]
```diff
@@ -336,3 +372,153 @@ def log_creation(self, *objects):
...
+            except ParserError:
+                continue
+
+        # make sure we have same number of users/dates by padding with None
+        while len(users) < len(dates):
+            users.insert(0, None)
+
+        # moving backwards in time, pair dates with users and event types
+        events = []
+        for i in range(len(dates) - 1, -1, -1):
```
Comment (rlskoeser) [inline geniza/corpus/management/commands/import_data.py:459]: This seems more complicated than it needs to be! Am having trouble following the logic.  Can't you just zip the two lists together? You can sort the dates independently to get the creation date if it isn't reliably first in the spreadsheet.

File: geniza/corpus/management/commands/import_data.py [hunk 4]
```diff
@@ -41,33 +47,57 @@
...
+    def setup(self, *args, **kwargs):
         if not hasattr(settings, 'DATA_IMPORT_URLS'):
-            raise CommandError('Please configure DATA_IMPORT_URLS in local settings')
+            raise CommandError(
+                'Please configure DATA_IMPORT_URLS in local settings')
 
+        # fetch users created through migrations & add to cache
         self.script_user = User.objects.get(username=settings.SCRIPT_USERNAME)
+        self.team_user = User.objects.get(username=settings.TEAM_USERNAME)
+        self.user_lookup["Geniza Lab team"] = self.team_user
```
[the remark this answers is not in the document]
Comment (rlskoeser) [reply geniza/corpus/management/commands/import_data.py:86]: 🤔 wonder if you could use a `defaultdict` where the lab team is the default value...

File: geniza/corpus/management/commands/import_data.py [hunk 5]
```diff
@@ -41,33 +47,57 @@
...
+
+        # load fixure containing known historic users (all non-active)
```
[the remark this answers is not in the document]
Comment (rlskoeser) [reply geniza/corpus/management/commands/import_data.py:88]: and/or also set verbosity to zero or to match current script verbosity

File: geniza/common/migrations/0005_create_staff_user.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
+# Generated by Django 3.1.6 on 2021-04-05 13:20
+
+from django.conf import settings
+from django.db import migrations
+
+
+def create_staff_user(apps, schema_editor):
```
Comment (rlskoeser) [inline geniza/common/migrations/0005_create_staff_user.py:7]: staff seems misleading/inaccurate for this, why not just call it geniza lab user?

File: geniza/corpus/management/commands/import_data.py [hunk 7]
```diff
@@ -336,3 +372,153 @@ def log_creation(self, *objects):
...
+        """Find a user account based on a provided name, using a simple cache.
+
+        If not found, tries to use first/last initials for lookup. If all else
+        fails, use the generic team account (TEAM_USERNAME).
+        """
+
+        # check the cache first
+        user = self.user_lookup.get(name)
+        if user:
+            logging.debug(
```
[the remark this answers is not in the document]
Comment (rlskoeser) [reply geniza/corpus/management/commands/import_data.py:386]: Thinking about it more — script output shouldn't require you to have a logger configured a particular way to see what's going on. (I manually edited my local copy to use `stdout.write` just so I could see some of what's going on.)

I know the log level handling is nice, so suggest you either configure a script-specific logger that will output things the right way without requiring any configuration or write command methods equivalent to the log levels that check the verbosity and output the message accordingly.

File: geniza/corpus/management/commands/import_data.py [hunk 8]
```diff
@@ -336,3 +372,153 @@ def log_creation(self, *objects):
...
+            event_type = "Created" if i == 0 else "Revised"
+
+            # if we have a date without a user, assign to the whole team
+            user = users[i] or self.team_user
+
+            # if it was two users together, log two separate events because
+            # each event must have exactly one user
+            if type(user) == tuple:
+                events.append(
+                    {"type": event_type, "user": user[0], "date": date})
```
Comment (rlskoeser) [inline geniza/corpus/management/commands/import_data.py:472]: use a named tuple?

File: geniza/corpus/tests/test_import_data.py [hunk 9]
```diff
@@ -362,6 +369,210 @@ def test_add_document_language():
...
+        "type": "Created",
+        "user": import_data_cmd.team_user,
+        "date": datetime.date(2016, 5, 4),
+    }]
+
+    # simplest case: one user, one well-formed date -> creation event
+    history = get_edit_history("Sarah Nisenson", "3/16/2018")
+    assert history == [{
+        "type": "Created",
+        "user": User.objects.get(username="snisenson"),
```
Comment (rlskoeser) [inline geniza/corpus/tests/test_import_data.py:451]: This is actually failing on the user comparison, not the date — different instances of the same django model must not be considered equal

It would be better to assert the comparisons separately so you can see where the error is when there's a mismatch. (And then you can just check username in the returned user instead of comparing to a different instance.)

**Review by rlskoeser — CHANGES_REQUESTED** [review]
Wow, this is a monster! Definitely need to upgrade the issue from 3 points to 5 if you haven't already.

Parts of this seem more complicated than necessary, and the reverse date looping is hard to follow — I'd like to see that revised. (Willing to work on it with you if we can find a time.)

I don't think it was ever documented in the issue, but as part of this work please remove the `old_input_by` and `old_input_date` fields from the `Document` model and from the import script.

File: geniza/corpus/management/commands/import_data.py [hunk 10]
```diff
@@ -41,33 +47,57 @@
...
+    2: logging.INFO,
+    3: logging.DEBUG
+}
+
 
 class Command(BaseCommand):
     'Import existing data from PGP spreadsheets into the database'
 
-    logentry_message = 'Created via data import script'
+    logentry_message = 'Imported via data import script'
```
[the remark this answers is not in the document]
Comment (thatbudakguy) [reply geniza/corpus/management/commands/import_data.py:66]: Can we just say "imported via script"? I changed it because other events will now sometimes say "created".

File: geniza/corpus/management/commands/import_data.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (thatbudakguy) [reply geniza/corpus/management/commands/import_data.py:86]: guess i don't actually need to add it to the lookup here, just saves one db query

File: geniza/corpus/management/commands/import_data.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (thatbudakguy) [reply geniza/corpus/management/commands/import_data.py:88]: nice, good idea to pass in the verbosity

File: geniza/corpus/management/commands/import_data.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (rlskoeser) [reply geniza/corpus/management/commands/import_data.py:66]: that works for me

File: geniza/corpus/management/commands/import_data.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (thatbudakguy) [reply geniza/corpus/management/commands/import_data.py:386]: I have the same problem, but I copied this implementation from the cdhweb exodus where it works without any logging config at all, so I don't know what's wrong. afaik this shouldn't interact with django's logging settings – it's just plain python logging using the root logger. Do you have any thoughts about why it might not be working?

File: geniza/corpus/management/commands/import_data.py [hunk 11]
```diff
@@ -336,3 +372,153 @@ def log_creation(self, *objects):
...
+            try:
+                user = User.objects.get(first_name__startswith=first_i,
+                                        last_name__startswith=last_i)
+            except User.DoesNotExist:
+                pass
+
+        # if we didn't get anyone through either method, warn and use team user
+        if not user:
+            logging.warning(
+                f"Couldn't find user {name} on PGPID {pgpid}; using {self.team_user}")
```
[the remark this answers is not in the document]
Comment (thatbudakguy) [reply geniza/corpus/management/commands/import_data.py:412]: That's because they're already in the database!

File: geniza/corpus/management/commands/import_data.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rlskoeser) [reply geniza/corpus/management/commands/import_data.py:86]: saving a db query is good! I just wasn't sure why you needed it in both the lookup and as a class variable

File: geniza/corpus/management/commands/import_data.py [hunk 12]
```diff
@@ -336,3 +372,153 @@ def log_creation(self, *objects):
...
+            return self.team_user
```
[the remark this answers is not in the document]
Comment (thatbudakguy) [reply geniza/corpus/management/commands/import_data.py:413]: It didn't seem simple to derive that from the spreadsheet they gave me – a lot of the entries were things I didn't need, and others I would have had to transform to make them usable. Overall there are only two ways to find a person (name or initials), so I thought this was the most straightforward way.

File: geniza/corpus/management/commands/import_data.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rlskoeser) [reply geniza/corpus/management/commands/import_data.py:386]: pretty sure cdhweb already has logging configured and example logging set up in the sample settings.

django _uses_ standard python logging, and python logging requires you to have logging configured to see the output

File: geniza/corpus/management/commands/import_data.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (rlskoeser) [reply geniza/corpus/management/commands/import_data.py:413]: yeah, I didn't expect you to generate it from the table they gave you; thought maybe you could generate from the fixture list once you had it, since you know the variants that occur

File: geniza/corpus/management/commands/import_data.py [hunk 13]
```diff
@@ -336,3 +372,153 @@ def log_creation(self, *objects):
...
+        # try to map every "input by" listing to a user account
+        users = []
+        for input_by in all_input:
+
+            # special case: two people together; add as a tuple
+            if " and " in input_by:
+                users.append(
+                    tuple(map(self.get_user, input_by.split(" and "))))
+            else:
+                users.append(self.get_user(input_by))
```
[the remark this answers is not in the document]
Comment (thatbudakguy) [reply geniza/corpus/management/commands/import_data.py:441]: not sure what you mean – I need to keep the ones paired with "and" together so that they can be matched to a single timestamp, because it's technically a single event with two authors

File: geniza/corpus/management/commands/import_data.py [hunk 14]
```diff
@@ -336,3 +372,153 @@ def log_creation(self, *objects):
...
+        dates = []
+        for date in all_dates:
+            try:
+                dates.append(parse(date, default=DEFAULT_EVENT_DATE).date())
+            except ParserError:
+                continue
+
+        # make sure we have same number of users/dates by padding with None
+        while len(users) < len(dates):
+            users.insert(0, None)
```
[the remark this answers is not in the document]
Comment (thatbudakguy) [reply geniza/corpus/management/commands/import_data.py:455]: there's a lot more going on here, let's discuss.

File: geniza/corpus/management/commands/import_data.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (rlskoeser) [reply geniza/corpus/management/commands/import_data.py:412]: They might be in QA, but they aren't in prod yet because this app doesn't exist there, and they weren't in my local dev database. I could see using `createcasuser` to make sure the user accounts exist rather than putting them in the fixture (which I agree seems a little weird), but it needs to be documented in deploy notes as a step to run (with any usernames that are required for data import) _before_ running the data import.

File: geniza/corpus/management/commands/import_data.py (hunk 1 above)
Comment (thatbudakguy) [reply geniza/corpus/management/commands/import_data.py:511]: in my mind, this function deals with how to represent the events created in the other function in django, so the logic of which flag matches which event belongs here rather than in the other function. that's also why the other function uses `date` instead of `datetime`.

File: geniza/corpus/management/commands/import_data.py [hunk 15]
```diff
@@ -336,3 +372,153 @@ def log_creation(self, *objects):
...
+
+        # finally, log the actual import event. if it's the only event, then
+        # it's an ADDITION – otherwise it's a CHANGE.
+        LogEntry.objects.log_action(
+            user_id=self.script_user.id,
+            content_type_id=self.content_types[Document].pk,
+            object_id=doc.pk,
+            object_repr=str(doc),
+            change_message=self.logentry_message,
+            action_flag=CHANGE if events else ADDITION
```
[the remark this answers is not in the document]
Comment (thatbudakguy) [reply geniza/corpus/management/commands/import_data.py:523]: as far as I can tell, django doesn't care! so no problem

File: requirements/prod.txt [hunk 16]
```diff
@@ -6,4 +6,5 @@ requests
 django-taggit
 #next version of piffle
 git+https://github.com/princeton-cdh/piffle.git@develop#egg=piffle
-taggit-selectize
\ No newline at end of file
+taggit-selectize
+python-dateutil
```
[the remark this answers is not in the document]
Comment (thatbudakguy) [reply requirements/prod.txt:10]: oh, I saw this too...I think some functions were renamed in the module. I'll check

File: geniza/corpus/management/commands/import_data.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (thatbudakguy) [reply geniza/corpus/management/commands/import_data.py:386]: in my cdhweb localsettings, there is no logging configured at all, and the exodus logging works just fine. I tried copying some different configurations into my geniza local dev, and none had any effect on the import_data command.

File: geniza/corpus/management/commands/import_data.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (thatbudakguy) [reply geniza/corpus/management/commands/import_data.py:412]: hmm, what do you think about creating them with `createcasuser` via `call_command` in this script? presumably it will do nothing if they already exist

[the remark this answers is not in the document]
Comment (rlskoeser) [reply geniza/corpus/management/commands/import_data.py:412]: even better! It will update the accounts from CAS if information has changed, but it won't cause any problems if they already exist. (You should go ahead and pass the admin flag also)

File: geniza/corpus/admin.py [hunk 17]
```diff
@@ -95,10 +95,9 @@ class DocumentAdmin(admin.ModelAdmin):
         'last_modified',
         'is_public'
     )
-    readonly_fields = ('old_input_by', 'old_input_date',
-                       'created', 'last_modified', 'shelfmark', 'id')
+    readonly_fields = ('created', 'last_modified', 'shelfmark', 'id')
     search_fields = ('fragments__shelfmark', 'tags__name', 'description',
-                     'old_input_by', 'notes', 'needs_review')
```
Comment (rlskoeser) [inline geniza/corpus/admin.py:101]: I hope they don't still need to search on this! Maybe we could make a list filter if it's important (although they have a lot already)

File: geniza/corpus/management/commands/import_data.py [hunk 18]
```diff
@@ -41,34 +48,72 @@
...
+        call_command("loaddata", "historic_users",
+                     app_label="corpus", verbosity=0)
+        logger.info("loaded 30 historic users")
+
+        # ensure current active users are present, but don't try to create them
+        # in a test environment because it's slow and requires VPN access
+        for username in ["rrichman", "mrustow", "ae5677", "alg4"]:
+            if "PYTEST_CURRENT_TEST" not in os.environ:
+                call_command("createcasuser", username,
+                             staff=True, verbosity=verbosity)
```
Comment (rlskoeser) [inline geniza/corpus/management/commands/import_data.py:100]: Why not run as a single command? That works on the command line, does it not work here, or is it slower?

File: geniza/corpus/management/commands/import_data.py [hunk 19]
```diff
@@ -345,3 +394,173 @@ def log_creation(self, *objects):
...
+        for input_by in all_input:
+
+            # for coauthored events, add both users to a tuple – otherwise
+            # it's a one-element tuple with the single user
+            if " and " in input_by:
+                first, second = input_by.split(" and ")
+                users.append((self.get_user(first, pgpid),
+                              self.get_user(second, pgpid)))
+            else:
+                users.append((self.get_user(input_by, pgpid),))
```
Comment (rlskoeser) [inline geniza/corpus/management/commands/import_data.py:470]: comment is out of date now (tuple)

this should be a one-liner — I think something like this:
```users.extend([self.get_users(u, pgpid) for u in input_by.split(" and ")])```

(you don't have to check for " and " in the string— you can just split on it and get a single-element list if it isn't present)

File: geniza/corpus/management/commands/import_data.py [hunk 20]
```diff
@@ -345,3 +394,173 @@ def log_creation(self, *objects):
...
+            users.insert(0, None)
+
+        # moving backwards in time, pair dates with users and event types.
+        # when there is a mismatch between number of users and dates, we want
+        # to associate the more recent dates with users, since in general
+        # we have more information the closer to the present we are. if we run
+        # out of users to assign, we use the generic team user.
+        events = []
+        users.reverse()
+        for i, date in enumerate(reversed(dates)):
```
Comment (rlskoeser) [inline geniza/corpus/management/commands/import_data.py:495]: much clearer, thank you

File: geniza/corpus/management/commands/import_data.py [hunk 21]
```diff
@@ -345,3 +394,173 @@ def log_creation(self, *objects):
...
+            else:
+                logger.debug(f"found event for PGPID {pgpid}: {events[-1]}")
+
+        # sort chronologically and return
+        events.sort(key=itemgetter("date"))
+        return events
+
+    sheet_add_msg = "Initial data entry (spreadsheet)"
+    sheet_chg_msg = "Major revision (spreadsheet)"
+
```
Comment (rlskoeser) [inline geniza/corpus/management/commands/import_data.py:520]: 👍

File: requirements/prod.txt [hunk 22]
```diff
@@ -9,4 +9,5 @@ git+https://github.com/princeton-cdh/piffle.git@develop#egg=piffle
 taggit-selectize
 django-multiselectfield
 django-modeltranslation
-django-admin-sortable2
\ No newline at end of file
+django-admin-sortable2
+python-dateutil>=2.8
```
Comment (rlskoeser) [inline requirements/prod.txt:13]: 👍

File: geniza/corpus/management/commands/import_data.py [hunk 23]
```diff
@@ -41,34 +48,72 @@
...
+        logger.setLevel(LOG_LEVELS[verbosity])
+
+        # load fixure containing known historic users (all non-active)
+        call_command("loaddata", "historic_users",
+                     app_label="corpus", verbosity=0)
+        logger.info("loaded 30 historic users")
+
+        # ensure current active users are present, but don't try to create them
+        # in a test environment because it's slow and requires VPN access
+        for username in ["rrichman", "mrustow", "ae5677", "alg4"]:
```
Comment (rlskoeser) [inline geniza/corpus/management/commands/import_data.py:97]: Given that this is slow and requires vpn, maybe we can skip it / only import for users that aren't already present? (if it's not too much trouble)

Here's an easy way to get the list of usernames already present:
```
User.objects.filter(username__in=["rrichman", "mrustow", "ae5677", "alg4"]).values_list('username', flat=True)
```

You can use a set difference or list filter to get the ones not in that list.

**Review by rlskoeser — APPROVED** [review]
Revisions look good, running locally and looks ok.

(Does this make the import a lot slower, or have I forgotten how slow it was already?)

File: geniza/corpus/management/commands/import_data.py (hunk 18 above)
Comment (thatbudakguy) [reply geniza/corpus/management/commands/import_data.py:100]: didn't even know you could! great suggestion

File: geniza/corpus/admin.py (hunk 17 above)
Comment (thatbudakguy) [reply geniza/corpus/admin.py:101]: how would that work if we're removing the field?

Comment (rlskoeser) [reply geniza/corpus/admin.py:101]: not on this field! a custom filter on users who have log entries for editing documents (but I hope we don't need this)

Comment (rlskoeser) [reply geniza/corpus/admin.py:101]: (when I said "this" I meant the information, not the field! sorry)

Comment (thatbudakguy) [reply geniza/corpus/admin.py:101]: oh, makes sense! yeah, I can see how that'd work but let's not do it unless we have to.
